File: BDFM-RAG-System/src/services/embedding.service.ts
import axios from 'axios';
import config from '../config';
import logger from '../utils/logger';
import { chunkText, sanitizeText } from '../utils/helpers';
import { VectorEmbedding, Correspondence, TextChunk } from '../models';
import { v4 as uuidv4 } from 'uuid';

export class EmbeddingService {
  private ollamaUrl: string;
  private model: string;

  constructor() {
    this.ollamaUrl = config.ollama.url;
    this.model = config.ollama.embeddingModel;
  }

  /**
   * Generate embedding for text
   */
  async generateEmbedding(text: string): Promise<number[]> {
    try {
      const sanitized = sanitizeText(text);

      const response = await axios.post(
        `${this.ollamaUrl}/api/embeddings`,
        {
          model: this.model,
          prompt: sanitized,
        },
        {
          timeout: config.ollama.timeout,
        }
      );

      if (!response.data || !response.data.embedding) {
        throw new Error('Invalid response from Ollama embedding API');
      }

      logger.debug(`Generated embedding for text (length: ${text.length})`);
      return response.data.embedding;
    } catch (error: any) {
      logger.error('Error generating embedding:', error.message);
      throw new Error(`Failed to generate embedding: ${error.message}`);
    }
  }

  /**
   * Generate embeddings for correspondence
   */
  async generateEmbeddingsForCorrespondence(
    correspondence: Correspondence
  ): Promise<VectorEmbedding[]> {
    try {
      logger.info(
        `Generating embeddings for correspondence ${correspondence.id}`
      );

      // Combine all text fields
      const fullText = this.buildCorrespondenceText(correspondence);

      // Chunk the text
      const chunks = chunkText(
        fullText,
        config.rag.chunkSize,
        config.rag.chunkOverlap
      );

      logger.info(
        `Split correspondence into ${chunks.length} chunks`
      );

      // Generate embeddings for each chunk
      const embeddings: VectorEmbedding[] = [];

      for (let i = 0; i < chunks.length; i++) {
        const chunk = chunks[i];
        const embeddingVector = await this.generateEmbedding(chunk);

        const embedding: VectorEmbedding = {
          id: uuidv4(),
          correspondenceId: correspondence.id,
          textChunk: chunk,
          embeddingVector,
          modelName: this.model,
          chunkIndex: i,
          language: this.detectLanguage(chunk),
          metadata: {
            correspondence_id: correspondence.id,
            mail_num: correspondence.mailNum,
            mail_date: correspondence.mailDate,
            subject: correspondence.subject,
            body_text: correspondence.bodyText,
            correspondence_type: correspondence.correspondenceType,
            secrecy_level: correspondence.secrecyLevel,
            priority_level: correspondence.priorityLevel,
            personality_level: correspondence.personalityLevel,
            created_at: correspondence.createdAt.toISOString(),
          },
          createdAt: new Date(),
        };

        embeddings.push(embedding);
      }

      logger.info(
        `Generated ${embeddings.length} embeddings for correspondence ${correspondence.id}`
      );

      return embeddings;
    } catch (error: any) {
      logger.error(
        `Error generating embeddings for correspondence ${correspondence.id}:`,
        error
      );
      throw error;
    }
  }

  /**
   * Generate embeddings for multiple correspondences
   */
  async generateBatchEmbeddings(
    correspondences: Correspondence[]
  ): Promise<Map<string, VectorEmbedding[]>> {
    const results = new Map<string, VectorEmbedding[]>();

    for (const correspondence of correspondences) {
      try {
        const embeddings =
          await this.generateEmbeddingsForCorrespondence(correspondence);
        results.set(correspondence.id, embeddings);
      } catch (error) {
        logger.error(
          `Failed to generate embeddings for correspondence ${correspondence.id}:`,
          error
        );
        // Continue with next correspondence
      }
    }

    return results;
  }

  /**
   * Build full text from correspondence
   */
  private buildCorrespondenceText(correspondence: Correspondence): string {
    const parts: string[] = [];

    // Add mail number and date
    parts.push(`رقم الكتاب: ${correspondence.mailNum}`);
    parts.push(`التاريخ: ${correspondence.mailDate}`);

    // Add subject
    if (correspondence.subject) {
      parts.push(`الموضوع: ${correspondence.subject}`);
    }

    // Add body
    if (correspondence.bodyText) {
      parts.push(`المحتوى: ${correspondence.bodyText}`);
    }

    // Add metadata
    parts.push(`نوع المراسلة: ${correspondence.correspondenceType}`);
    parts.push(`مستوى الأولوية: ${correspondence.priorityLevel}`);
    parts.push(`مستوى السرية: ${correspondence.secrecyLevel}`);

    // Add external reference if exists
    if (correspondence.externalReferenceNumber) {
      parts.push(
        `المرجع الخارجي: ${correspondence.externalReferenceNumber}`
      );
    }

    return parts.join('\n\n');
  }

  /**
   * Detect language of text (simple heuristic)
   */
  private detectLanguage(text: string): string {
    // Check for Arabic characters
    const arabicRegex = /[\u0600-\u06FF]/;
    if (arabicRegex.test(text)) {
      return 'ar';
    }
    return 'en';
  }

  /**
   * Check if Ollama is available
   */
  async checkConnection(): Promise<boolean> {
    try {
      const response = await axios.get(`${this.ollamaUrl}/api/tags`, {
        timeout: 5000,
      });
      return response.status === 200;
    } catch (error) {
      logger.error('Ollama connection check failed:', error);
      return false;
    }
  }

  /**
   * Get available models
   */
  async getAvailableModels(): Promise<string[]> {
    try {
      const response = await axios.get(`${this.ollamaUrl}/api/tags`, {
        timeout: 5000,
      });

      if (response.data && response.data.models) {
        return response.data.models.map((m: any) => m.name);
      }

      return [];
    } catch (error) {
      logger.error('Error getting available models:', error);
      return [];
    }
  }

  /**
   * Verify embedding model is available
   */
  async verifyModel(): Promise<boolean> {
    try {
      const models = await this.getAvailableModels();
      const isAvailable = models.includes(this.model);

      if (!isAvailable) {
        logger.warn(
          `Embedding model ${this.model} is not available. Available models: ${models.join(', ')}`
        );
      }

      return isAvailable;
    } catch (error) {
      logger.error('Error verifying embedding model:', error);
      return false;
    }
  }
}

// Export singleton instance
export default new EmbeddingService();

File: BDFM-RAG-System/bdfm-rag-python/app/services/embedding_service.py
"""Embedding service using sentence-transformers with multilingual-e5-large"""
from sentence_transformers import SentenceTransformer
import numpy as np
from typing import List, Dict, Any
import asyncio
from app.config import settings
from app.utils.logger import logger
from app.utils.helpers import chunk_text, generate_deterministic_id


class EmbeddingService:
    """Embedding service for generating vector embeddings using multilingual-e5-large"""
    
    def __init__(self, model_name: str = None):
        self.model_name = model_name or settings.embedding_model
        self.model: SentenceTransformer = None
        self.dimension = settings.embedding_dimension
        self._loading = False
        
    def initialize(self):
        """تحميل النموذج (Lazy loading)"""
        if self.model is not None or self._loading:
            return
            
        self._loading = True
        logger.info(f"Loading embedding model: {self.model_name}")
        
        try:
            self.model = SentenceTransformer(self.model_name)
            
            # Test embedding to get actual dimension
            test_embedding = self.model.encode("test", show_progress_bar=False)
            actual_dimension = len(test_embedding)
            
            if actual_dimension != self.dimension:
                logger.warning(
                    f"Expected dimension {self.dimension} but got {actual_dimension}. Updating config."
                )
                self.dimension = actual_dimension
            
            logger.info(f"Model loaded successfully. Dimension: {self.dimension}")
            
        except Exception as e:
            logger.error(f"Error loading embedding model: {e}")
            raise
        finally:
            self._loading = False
    
    async def generate_embedding(self, text: str, is_query: bool = False) -> List[float]:
        """
        توليد embedding لنص واحد
        
        Args:
            text: النص المراد ترميزه
            is_query: هل النص استعلام (query) أم passage
            
        Returns:
            Embedding vector كـ list
        """
        if not self.model:
            # تحميل في executor لتجنب blocking event loop
            loop = asyncio.get_event_loop()
            await loop.run_in_executor(None, self.initialize)
        
        # تشغيل encoding في executor
        loop = asyncio.get_event_loop()
        embedding = await loop.run_in_executor(
            None,
            self._encode_text,
            text,
            is_query
        )
        
        return embedding.tolist()
    
    def _encode_text(self, text: str, is_query: bool = False) -> np.ndarray:
        """
        ترميز النص (يعمل في thread منفصل)
        
        Args:
            text: النص
            is_query: هل هو query أم passage
            
        Returns:
            NumPy array of embedding
        """
        # E5 models تحتاج prefix
        # للـ queries: "query: ..."
        # للـ passages: "passage: ..."
        if 'e5' in self.model_name.lower():
            prefix = "query:" if is_query else "passage:"
            text = f"{prefix} {text}"
        
        embedding = self.model.encode(
            text,
            normalize_embeddings=True,
            show_progress_bar=False,
            convert_to_numpy=True
        )
        
        return embedding
    
    async def generate_embeddings_batch(
        self,
        texts: List[str],
        batch_size: int = 32,
        is_query: bool = False
    ) -> List[List[float]]:
        """
        توليد embeddings لعدة نصوص مع batch processing
        
        Args:
            texts: قائمة النصوص
            batch_size: حجم الدفعة
            is_query: هل النصوص queries أم passages
            
        Returns:
            قائمة من embeddings
        """
        if not self.model:
            loop = asyncio.get_event_loop()
            await loop.run_in_executor(None, self.initialize)
        
        loop = asyncio.get_event_loop()
        embeddings = await loop.run_in_executor(
            None,
            self._encode_batch,
            texts,
            batch_size,
            is_query
        )
        
        return embeddings.tolist()
    
    def _encode_batch(
        self,
        texts: List[str],
        batch_size: int,
        is_query: bool = False
    ) -> np.ndarray:
        """
        ترميز دفعة من النصوص
        
        Args:
            texts: النصوص
            batch_size: حجم الدفعة
            is_query: نوع النصوص
            
        Returns:
            NumPy array of embeddings
        """
        # إضافة prefix للـ E5
        if 'e5' in self.model_name.lower():
            prefix = "query:" if is_query else "passage:"
            texts = [f"{prefix} {text}" for text in texts]
        
        embeddings = self.model.encode(
            texts,
            batch_size=batch_size,
            normalize_embeddings=True,
            show_progress_bar=True,
            convert_to_numpy=True
        )
        
        return embeddings
    
    async def generate_embeddings_for_correspondence(
        self,
        correspondence: Dict[str, Any]
    ) -> List[Dict[str, Any]]:
        """
        توليد embeddings لمراسلة مع chunking
        
        Args:
            correspondence: بيانات المراسلة
            
        Returns:
            قائمة من embeddings مع metadata
        """
        # بناء النص الكامل
        full_text = self._build_correspondence_text(correspondence)
        
        # تقسيم إلى chunks
        chunks = chunk_text(
            full_text,
            chunk_size=settings.chunk_size,
            chunk_overlap=settings.chunk_overlap
        )
        
        logger.info(
            f"Generated {len(chunks)} chunks for correspondence {correspondence.get('id')}"
        )
        
        # توليد embeddings لكل chunk
        embeddings_list = await self.generate_embeddings_batch(
            chunks,
            batch_size=32,
            is_query=False  # Passages
        )
        
        # بناء النتيجة
        results = []
        for i, (chunk, embedding) in enumerate(zip(chunks, embeddings_list)):
            # استخدام ID حتمي
            embedding_id = generate_deterministic_id(correspondence.get('id'), i)
            
            results.append({
                'id': embedding_id,
                'correspondence_id': correspondence.get('id'),
                'text_chunk': chunk,
                'embedding_vector': embedding,
                'chunk_index': i,
                'metadata': {
                    'mail_num': correspondence.get('mail_num'),
                    'mail_date': correspondence.get('mail_date'),
                    'subject': correspondence.get('subject'),
                    'correspondence_type': correspondence.get('correspondence_type'),
                    'secrecy_level': correspondence.get('secrecy_level'),
                    'priority_level': correspondence.get('priority_level'),
                    'personality_level': correspondence.get('personality_level'),
                    'created_at': correspondence.get('created_at').isoformat() if correspondence.get('created_at') else None,
                }
            })
        
        return results
    
    def _build_correspondence_text(self, correspondence: Dict[str, Any]) -> str:
        """
        بناء النص الكامل من المراسلة
        
        Args:
            correspondence: بيانات المراسلة
            
        Returns:
            النص الكامل
        """
        parts = []
        
        parts.append(f"رقم الكتاب: {correspondence.get('mail_num', '')}")
        parts.append(f"التاريخ: {correspondence.get('mail_date', '')}")
        
        if correspondence.get('subject'):
            parts.append(f"الموضوع: {correspondence['subject']}")
        
        if correspondence.get('body_text'):
            parts.append(f"المحتوى: {correspondence['body_text']}")
        
        parts.append(f"نوع المراسلة: {correspondence.get('correspondence_type', '')}")
        parts.append(f"مستوى الأولوية: {correspondence.get('priority_level', '')}")
        parts.append(f"مستوى السرية: {correspondence.get('secrecy_level', '')}")
        
        if correspondence.get('external_reference_number'):
            parts.append(f"المرجع الخارجي: {correspondence['external_reference_number']}")
        
        return '\n\n'.join(parts)
    
    def get_dimension(self) -> int:
        """الحصول على dimension"""
        return self.dimension
    
    async def check_model_loaded(self) -> bool:
        """فحص إذا كان النموذج محمل"""
        return self.model is not None


# Create global embedding service instance
embedding_service = EmbeddingService()


__all__ = ["embedding_service", "EmbeddingService"]

